src/analysis/special_functions/trigonometric.lean

```diff
@@ -1690,4 +1665,84 @@ by cases n; simp only [cos_nat_mul_two_pi, int.of_nat_eq_coe,
 lemma cos_int_mul_two_pi_add_pi (n : ℤ) : cos (n * (2 * π) + π) = -1 :=
 by simp [cos_add, sin_add, cos_int_mul_two_pi]
 
+lemma exp_pi_mul_I : exp (π * I) = -1 := by { rw exp_mul_I, simp, }
+
+theorem cos_eq_zero_iff {θ : ℂ} : cos θ = 0 ↔ ∃ k : ℤ, θ = (2 * k + 1) * π / 2 :=
+begin
+  have h : (exp (θ * I) + exp (-θ * I)) / 2 = 0 ↔ exp (2 * θ * I) = -1,
+  { rw [@div_eq_iff _ _ (exp (θ * I) + exp (-θ * I)) 2 0 (by norm_num), zero_mul, add_eq_zero_iff_eq_neg, neg_eq_neg_one_mul (exp (-θ * I)), ← div_eq_iff (exp_ne_zero (-θ * I)), ← exp_sub],
+    field_simp, ring },
+  rw [cos, h, ← exp_pi_mul_I, exp_eq_exp_iff_exists_int],
+  split; simp; intros; use x,
+  { field_simp, ring at a,
+    rwa [mul_right_comm 2 I θ, mul_right_comm (2*(x:ℂ)+1) I (π:ℂ), mul_left_inj' I_ne_zero, mul_comm 2 θ] at a },
+  { field_simp at a, ring,
+    rw [mul_right_comm 2 I θ, mul_right_comm (2*(x:ℂ)+1) I (π:ℂ), mul_left_inj' I_ne_zero, mul_comm 2 θ, a] },
+end
+
+theorem cos_ne_zero_iff {θ : ℂ} : cos θ ≠ 0 ↔ ∀ k : ℤ, θ ≠ (2 * k + 1) * π / 2 :=
+by rw [← not_exists, not_iff_not, cos_eq_zero_iff]
+
+lemma has_deriv_at_tan {x:ℂ} (h : ∀ k : ℤ, x ≠ (2 * k + 1) * π / 2) : has_deriv_at tan (1 / (cos x)^2) x :=
+begin
+  convert has_deriv_at.div (has_deriv_at_sin x) (has_deriv_at_cos x) (cos_ne_zero_iff.mpr h),
+  rw ← sin_sq_add_cos_sq x,
+  ring,
+end
+
+lemma differentiable_at_tan {x:ℂ} (h : ∀ k : ℤ, x ≠ (2 * k + 1) * π / 2) : differentiable_at ℂ tan x :=
+(has_deriv_at_tan h).differentiable_at
+
+lemma deriv_tan {x:ℂ} (h : ∀ k : ℤ, x ≠ (2 * k + 1) * π / 2): deriv tan x = 1 / (cos x)^2 :=
+(has_deriv_at_tan h).deriv
+
+lemma continuous_tan : continuous (λ x : {x // cos x ≠ 0}, tan x) :=
+(continuous_sin.comp continuous_subtype_val).mul
+  (continuous.inv subtype.property (continuous_cos.comp continuous_subtype_val))
+
 end complex
+
+namespace real
+
+open_locale real
+
+theorem cos_eq_zero_iff {θ : ℝ} : cos θ = 0 ↔ ∃ k : ℤ, θ = (2 * k + 1) * π / 2 :=
+begin
+  rw [← complex.of_real_eq_zero, complex.of_real_cos θ],
+  convert @complex.cos_eq_zero_iff θ,
+  norm_cast,
+end
+
+theorem cos_ne_zero_iff {θ : ℝ} : cos θ ≠ 0 ↔ ∀ k : ℤ, θ ≠ (2 * k + 1) * π / 2 :=
+by rw [← not_exists, not_iff_not, cos_eq_zero_iff]
+
+
+lemma has_deriv_at_tan {x:ℝ} (h : ∀ k : ℤ, x ≠ (2 * k + 1) * π / 2) : has_deriv_at tan (1 / (cos x)^2) x :=
+begin
+  convert has_deriv_at_real_of_complex (complex.has_deriv_at_tan (by { convert h, norm_cast } )),
+  rw ← complex.of_real_re (1/((cos x)^2)),
+  simp,
+end
+
+lemma differentiable_at_tan {x:ℝ} (h : ∀ k : ℤ, x ≠ (2 * k + 1) * π / 2) : differentiable_at ℝ tan x :=
+(has_deriv_at_tan h).differentiable_at
+
+lemma deriv_tan {x:ℝ} (h : ∀ k : ℤ, x ≠ (2 * k + 1) * π / 2) : deriv tan x = 1 / (cos x)^2 :=
+(has_deriv_at_tan h).deriv
+
+lemma continuous_tan : continuous (λ x : {x // cos x ≠ 0}, tan x) :=
```

Comment:
Same remark about `continuous_on`.

Reply:
Again, I did not add this lemma, it was already in the file the way that it is